hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/README.gRPC.md

```diff
@@ -0,0 +1,253 @@
+guarantee such an ordering. More details about the reasons will be shared later
+in this doc. In a nutshell, it brings in performance and complexity penalties 
+with no real gains.
+
+Retries are not handled nor provided for async calls, for these calls, the
+client code has to implement the retry logic on its own, as that may involve
+re-sending multiple requests, and may affect request ordering as well.
+
+
+## How clients can be created
+The `XCeiverClientManager` class is managing creation of the client instances,
+as defined in the `XCeiverClientFactory` interface.
+A user of this class can acquire a client with the help of acquire methods, and
+then release it when it is not using it anymore.
+
+It is important to note that the lifecycle of the clients are up to the user,
+and if a client is not released it may leak, as the clients are reference
+counted and evicted from the cache if they are not references for more than
+`scm.container.client.idle.threshold` milliseconds ago. The default idle
+threshold is 10 seconds.
+
+The client is getting connected to the DataNode upon creation, and at every
+request it checks if the connection is still up and ready. In case the
+connection was terminated for any reason, the client simply tries to reconnect
+to the DataNode.
+
+
+## Client to server communication in a nutshell
+In gRPC there are 3 things that is part of the communication, stub, channel, and
+messages.
+
+Stub provides tha server side API on the client side.
+Channel manages the communication properties, and the socket level connection.
+Messages in our case are sent via bi-directional streaming with the help of
+StreamObservers.
+
+When we initiate a connection we create a stub, and a channel, and when we send
+a request (a message) we create the bi-directional stream, we send in a 
+request proto message, and we get back a response proto message, then we close
+the stream.
+We never send multiple requests via the same stream, although we can if we want
+to, there are reasons that will be discussed with more details later on.
+
+The server side of communication processes the requests in the same thread in
+which they arrive, via `HDDSDispatcher` in `GrpcXceiverService`, and this also
+means that the requests are processed in more threads as the gRPC server side
+provides multi-threaded processing per streams.
+
+
+## Client to server communication in more detail
+
+### How requests are handled on client and server side
+The gRPC client is also known as the standalone client, is used only for
+reads in the current environment, but during the Erasure Coded storage
+development we already selected the standalone client for writes. That effort
+initiated the analysis of the client side, and this writeup.
+The plan is to continue to use the gRPC client only for reads, and for EC writes
+in the future, if that changes, then this document might as well need an update.
```

Comment:
It is very easy to miss to update this document if we change our mind to use some other ways that GrpcClient for EC writes. If we expect this doc to be updated, we may need to provide back reference in relevant java file.

Reply:
I prefer this kind of detailed docs can be at javadocs also, but this looks to be having too many details.

Reply:
In #3012 I have referred the file from the class' API doc, and also I removed the sentence starting as "The plan is..." this plan is documented in the JIRA tasks, and will be visible as the code evolves, when changes to be made, hopefully one does not forget to update this doc, if it is necessary.